### PR TITLE
feat(cli): allow configuration of http-server's timeout configuration

### LINKED ISSUE #28633: Please expose keepAliveTimeout and headersTimeout without requiring a custom server
### Describe the feature you'd like to request

When deploying behind a load balancer its good to configure the application server with a longer timeout than the load balancer. See [AWS ALB reccomendation](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/application-load-balancers.html#connection-idle-timeout) 

> We also recommend that you configure the idle timeout of your application to be larger than the idle timeout configured for the load balancer. 

AWS defaults a ALB to a 60 second idle timeout so it would be desirable to configure next to have a 61 second timeout.

It looks like this has been discussed in a few places with little to no response from the maintainers:
- https://github.com/vercel/next.js/discussions/16544
- https://github.com/vercel/next.js/discussions/19618
- https://github.com/vercel/next.js/issues/22204

Here is a discussion of a similar issue in node. https://github.com/nodejs/node/issues/27363
and two blogs about people hitting similar issues with ALB's generating 502s using express servers where the solution was to increase the keepAliveTimeout.
- https://adamcrowder.net/posts/node-express-api-and-aws-alb-502/
- https://shuheikagawa.com/blog/2019/04/25/keep-alive-timeout/

### Describe the solution you'd like

Allow setting keepAliveTimeout and headersTimeout.

### Describe alternatives you've considered

Ugly hacks 

> We are doing something similar to patch, just an awk on the bundle after build. As patch replaces the whole file (IIRC) it’s more likely to break things than a regex search and insert. 

https://github.com/vercel/next.js/discussions/19618#discussioncomment-1218511

> We're using patch-package to fix this issue, please make server.keepAliveTimeout a next.js settings 🙏🏻

https://github.com/vercel/next.js/discussions/19618#discussioncomment-1218160

or setting the ALB/ELB to a timeout of 4 seconds
> Heya the only solution I was able to find was to modify the IdleTimeout on the AWS ELB to be 4 seconds, i.e 1 second less than node's default 5 seconds.

https://github.com/vercel/next.js/discussions/16544#discussioncomment-142087

### PR DESCRIPTION
**Pull request by Miikis**
## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

## Feature

- [x] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [x] Related issues linked using `fixes #number`
- [x] Integration tests added
- [x] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [x] Make sure the linting passes by running `yarn lint`

fixes #28633

## Related Discussions
- https://github.com/vercel/next.js/discussions/22211
- https://github.com/vercel/next.js/discussions/28642
- https://github.com/vercel/next.js/discussions/19618

### CODE REVIEW

**Comment by Miikis** [comment]
Couldn't find the tests for `next start` on first-pass. Point me to where they are and I'll add some tests. Let me know if the PR needs more things. Thanks 😊

**Comment by tvelingkar** [comment]
@timneutkens When can we expect this to be merged?

File: docs/api-reference/cli.md
Comment (styfle) [inline docs/api-reference/cli.md:111]: All mentions of "upstream" in this description should be "downstream".

- Next.js is "upstream"
- Proxy is "downstream"

File: docs/api-reference/cli.md
Comment (styfle) [inline docs/api-reference/cli.md:122]: Should this be configured in `next.config.js` instead?

https://nextjs.org/docs/api-reference/next.config.js/disabling-http-keep-alive

File: packages/next/cli/next-start.ts
Comment (styfle) [inline packages/next/cli/next-start.ts:67]: This needs to have error handling when the user passes a non-integer

Also needs tests

File: docs/api-reference/cli.md
Comment (Miikis) [reply docs/api-reference/cli.md:122]: I thought about it but in [this thread](https://github.com/vercel/next.js/discussions/28642#discussioncomment-1805011) someone asked if these settings could be environment variables and @timneutkens made a good point about how that wouldn't make sense — since the changes only apply to `next start`. I just assumed the same logic would apply to `next.config.js`. Let me know if you think different.

File: packages/next/cli/next-start.ts
Comment (Miikis) [reply packages/next/cli/next-start.ts:67]: Should it throw or swallow the error? I figured an invalid value should throw so the user would know immediately. And what would throw there is a TypeError saying something about NaN. How do you think it should be handled?

File: docs/api-reference/cli.md
Comment (mAAdhaTTah) [inline docs/api-reference/cli.md:111]: I think it's actually the opposite: Next.js should be configured with a timeout that is _larger_ than the LB's timeout. See here: https://adamcrowder.net/posts/node-express-api-and-aws-alb-502/ We had 502s and set our timeout slightly higher than the LBs and that resolved them for us.

Comment (mAAdhaTTah) [reply docs/api-reference/cli.md:111]: See also here: https://docs.aws.amazon.com/elasticloadbalancing/latest/application/application-load-balancers.html#connection-idle-timeout

> We also recommend that you configure the idle timeout of your application to be larger than the idle timeout configured for the load balancer. Otherwise, if the application closes the TCP connection to the load balancer ungracefully, the load balancer might send a request to the application before it receives the packet indicating that the connection is closed. If this is the case, the server refuses the request from the load balancer and then the load balancer sends an HTTP 502 Bad Gateway error to the application.

Comment (Miikis) [reply docs/api-reference/cli.md:111]: yea, you're right lol that's actually what i meant

Comment (mAAdhaTTah) [reply docs/api-reference/cli.md:111]: FWIW, I _kinda_ disagree with this suggestion. If the request comes into the LB initially, then Next.js is "downstream" from it.

Comment (Miikis) [reply docs/api-reference/cli.md:111]: yea, it's a matter of perspective. i'm okay going either way — if you don't feel too strongly about it, I'm just going to go w/ @styfle suggestion :)

File: packages/next/cli/next-start.ts
Comment (Miikis) [reply packages/next/cli/next-start.ts:67]: I'm checking for NaN for now and exiting if invalid.

File: test/integration/cli/test/index.test.js
Comment (Miikis) [inline test/integration/cli/test/index.test.js:398]: Added these integration tests but locally they hang. Not sure why.

File: docs/api-reference/cli.md
Comment (styfle) [inline docs/api-reference/cli.md:119]: > In some earlier versions of Node.js

In which earlier versions? It looks like Node.js 10 and earlier but Next.js only supports 12+ (soon to be 14+) so I think we don't need this setting.

Comment (Miikis) [reply docs/api-reference/cli.md:119]: The [fix doesn't appear until 12.19](https://github.com/nodejs/node/issues/27363#issuecomment-705085158) it looks like. Let's keep this setting until Next.js drops support for 12?

File: test/integration/cli/test/index.test.js
Comment (Miikis) [reply test/integration/cli/test/index.test.js:398]: @styfle can you approve the workflows so we can see if these fail in the ci? Or do you have some insight as to why they're hanging when ran locally?

File: docs/api-reference/cli.md
Comment (styfle) [reply docs/api-reference/cli.md:119]: Node.js 12 reaches EOL April 30th. We might not even have a stable release of Next.js before then. I don't think the additional flags and docs are worth it.

**Comment by ijjk** [comment]
## Failing test suites 
Commit: ab9f2ba22b83891dc7d14d1513107f0d1a2bfd93

`yarn testheadless test/integration/cli/test/index.test.js` 
- CLI Usage > info > should print output
- CLI Usage > start > should validate keepAliveTimeout argument
- CLI Usage > start > should validate headersTimeout argument

<details>
<summary>Expand output</summary>

  ● CLI Usage › start › should validate keepAliveTimeout argument

    expect(received).toEqual(expected) // deep equality

    - Expected  -  1
    + Received  + 12

    - Invalid type for --keepAliveTimeout; provide an integer in milliseconds
    + Error: listen EADDRINUSE: address already in use 0.0.0.0:3000
    +     at Server.setupListenHandle [as _listen2] (net.js:1331:16)
    +     at listenInCluster (net.js:1379:12)
    +     at doListen (net.js:1516:7)
    +     at processTicksAndRejections (internal/process/task_queues.js:83:21) {
    +   code: 'EADDRINUSE',
    +   errno: -98,
    +   syscall: 'listen',
    +   address: '0.0.0.0',
    +   port: 3000
    + }
    +

      403 |         }
      404 |       )
    > 405 |       expect(output.stderr).toEqual(
          |                             ^
      406 |         'Invalid type for --keepAliveTimeout; provide an integer in milliseconds'
      407 |       )
      408 |     })

      at Object.<anonymous> (integration/cli/test/index.test.js:405:29)

  ● CLI Usage › start › should validate headersTimeout argument

    expect(received).toEqual(expected) // deep equality

    - Expected  -  1
    + Received  + 12

    - Invalid type for --headersTimeout; provide an integer in milliseconds
    + Error: listen EADDRINUSE: address already in use 0.0.0.0:3000
    +     at Server.setupListenHandle [as _listen2] (net.js:1331:16)
    +     at listenInCluster (net.js:1379:12)
    +     at doListen (net.js:1516:7)
    +     at processTicksAndRejections (internal/process/task_queues.js:83:21) {
    +   code: 'EADDRINUSE',
    +   errno: -98,
    +   syscall: 'listen',
    +   address: '0.0.0.0',
    +   port: 3000
    + }
    +

      415 |         }
      416 |       )
    > 417 |       expect(output.stderr).toEqual(
          |                             ^
      418 |         'Invalid type for --headersTimeout; provide an integer in milliseconds'
      419 |       )
      420 |     })

      at Object.<anonymous> (integration/cli/test/index.test.js:417:29)

  ● CLI Usage › info › should print output

    expect(received).toBe(expected) // Object.is equality

    Expected: ""
    Received: "warn  - Latest canary version not detected, detected: \"12.1.5-canary.2\", newest: \"12.1.5-canary.4\".
            Please try the latest canary version (`npm install next@canary`) to confirm the issue still exists before creating a new issue.
            Read more - https://nextjs.org/docs/messages/opening-an-issue
    "

      520 |       // warning will show so skip this check for the stable release
      521 |       if (pkg.version.includes('-canary')) {
    > 522 |         expect(info.stderr || '').toBe('')
          |                                   ^
      523 |       }
      524 |       expect(info.stdout).toMatch(
      525 |         new RegExp(`

      at Object.<anonymous> (integration/cli/test/index.test.js:522:35)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | Miikis/next.js feat/keep-alive-timeouts | Change |
| - | - | - | - |
| buildDuration | 18.4s | 18.2s | -266ms |
| buildDurationCached | 7.3s | 7.2s | -95ms |
| nodeModulesSize | 477 MB | 478 MB | ⚠️ +1.97 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | Miikis/next.js feat/keep-alive-timeouts | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.825 | 3.79 | -0.04 |
| / avg req/sec | 653.62 | 659.65 | +6.03 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.58 | 1.606 | ⚠️ +0.03 |
| /error-in-render avg req/sec | 1582.41 | 1556.2 | ⚠️ -26.21 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | Miikis/next.js feat/keep-alive-timeouts | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28 kB | 28 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.7 kB | 71.7 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | Miikis/next.js feat/keep-alive-timeouts | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | Miikis/next.js feat/keep-alive-timeouts | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 309 B | 309 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 3.05 kB | 3.05 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.64 kB | 5.64 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.32 kB | 2.32 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 387 B | 387 B | ✓ |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.9 kB | 15.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | Miikis/next.js feat/keep-alive-timeouts | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | Miikis/next.js feat/keep-alive-timeouts | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 544 B | 544 B | ✓ |
| withRouter.html gzip | 524 B | 524 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | Miikis/next.js feat/keep-alive-timeouts | Change |
| - | - | - | - |
| buildDuration | 21s | 20.8s | -215ms |
| buildDurationCached | 7.2s | 7.1s | -101ms |
| nodeModulesSize | 477 MB | 478 MB | ⚠️ +1.97 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | Miikis/next.js feat/keep-alive-timeouts | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.75 | 3.798 | ⚠️ +0.05 |
| / avg req/sec | 666.71 | 658.2 | ⚠️ -8.51 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.607 | 1.633 | ⚠️ +0.03 |
| /error-in-render avg req/sec | 1555.21 | 1531.33 | ⚠️ -23.88 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | Miikis/next.js feat/keep-alive-timeouts | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.3 kB | 28.3 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.2 kB | 72.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | Miikis/next.js feat/keep-alive-timeouts | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | Miikis/next.js feat/keep-alive-timeouts | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 313 B | 313 B | ✓ |
| css-HASH.js gzip | 325 B | 325 B | ✓ |
| dynamic-HASH.js gzip | 3.03 kB | 3.03 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.7 kB | 5.7 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 388 B | 388 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.9 kB | 15.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | Miikis/next.js feat/keep-alive-timeouts | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | Miikis/next.js feat/keep-alive-timeouts | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 544 B | 544 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: ab9f2ba22b83891dc7d14d1513107f0d1a2bfd93

File: docs/api-reference/cli.md
Comment (Miikis) [reply docs/api-reference/cli.md]: headersTimeout is removed

File: test/integration/cli/test/index.test.js
Comment (Miikis) [reply test/integration/cli/test/index.test.js]: This test was unnecessary because the `args` module seems to already filter for `NaN`s.

File: packages/next/cli/next-start.ts
Comment (Miikis) [inline packages/next/cli/next-start.ts:66]: I removed the input-type validation (`args` module seems to already take care of this) and replaced it with some input-range validation. Then input-range validation clamps the `--keepAliveTimeout` value to be between 0 and Infinity-1, inclusive. 

There doesn't seem to be a way to write a test for this since 1) the invalid-range validation will not throw errors and 2) the underlying `http` server is not exposed from `startServer`. If the lack of tests for this validation will hold up the PR, I can revert the commit that introduced the range-validation. Let me know.

It would be nice to get this PR in in the next canary release. Let me know if you guys need something more, @styfle  @timneutkens

**Comment by Miikis** [comment]
@styfle could you let the workflows run again in the meantime?

File: packages/next/cli/next-start.ts
Comment (styfle) [reply packages/next/cli/next-start.ts:66]: Can you add a test to confirm it works properly and won't regress in the future?

**Comment by devcitopia** [comment]
Any update on this ?
Can we give a hand to get this merged ?

File: packages/next/cli/next-start.ts
Comment (styfle) [inline packages/next/cli/next-start.ts:71]:
```suggestion
      `Invalid `--keepAliveTimeout`, expected a non negative number but received ${keepAliveTimeoutArg}`,
```

**Review by styfle — DISMISSED** [review]
Thanks!

**Review by balazsorban44 — DISMISSED** [review]

**Review by styfle — APPROVED** [review]

**Review by balazsorban44 — APPROVED** [review]

**Pull request merged**
